Logic: Add `canonical` property to `Not`

mlange05 opened the pull request:
#### Brief description of what is fixed or changed

The `_new_args_filter` for `And` and `Or` calls `(~c).canonical` explicitly,
which cause `AttributeError`s because that property is not implemented.

To trigger the error, try the following snippet:
```
from sympy import *
a, b, c, d = symbols('a b c d')
expr = (a > b) & (b > c)

with evaluate(False):
    print(expr.xreplace({b: d}))
```

#### Other comments
I'm not 100% that this is the preferred way to fix this, but it highlights the issue and contains one very simple solution. If there is a better way to prevent the given example from triggering an `AttributeError`, please let me know.

<!-- BEGIN RELEASE NOTES -->
* logic
  * Added canonical property to Not
<!-- END RELEASE NOTES -->

mlange05 committed:
Logic: Add `caconical` property to `Not`

The `_new_args_filter` for `And` and `Or` calls this explicitly,
which cause `AttributeError`s. To trigger, try the following snippet:
```
from sympy import *
a, b, c, d = symbols('a b c d')
expr = (a > b) & (b > c)

with evaluate(False):
    print(expr.xreplace({b: d}))
```

oscargus commented:
A slightly simpler way to obtain the error is:
```
from sympy import *
a, b = symbols('a b')
c = a > 1
d = And(c, Not(c))
```
(Well, more lines, but it is not required to do `xreplace`)

It is actually quite surprising that the current tests pass...

mlange05 committed:
Logic: Add a small test for the existence of `Not.canonical`

mlange05 commented:
@oscargus Thanks for the prompt reply. I've added two small tests, one that basically test the existence of the property `canonical` on `Not` objects and one that tests the behaviour of `Not(A > 1).canonical`. Please note that the second test would actually pass on `master`, since the simplification transforms `Not(A > 1)` into `A <= 1`, unless `evaluate(False)` is used, as in my previous example. However, this should now be caught by the `Not(A).canonical` test.

smichr commented on sympy/logic/boolalg.py:
```diff
@@ -736,6 +736,10 @@ def eval(cls, arg):
 
+    @property
+    def canonical(self):
+        return self
```
I would suggest `return self.func(*Eq(*self.args).canonical.args)` so whatever is done to make Eq canonical will be done for Ne.

smichr commented on sympy/logic/tests/test_boolalg.py:
```diff
@@ -114,6 +114,9 @@ def test_Not():
     assert Not(2) is false
+    assert Not(A).canonical == Not(A)
+    e = A > 1
+    assert Not(e).canonical == Not(e)
```
`Not(e)` is automatically converted to `A <=1` and so this does not test the added code; try `Not(A).canonical`.

mlange05 replied:
@smichr I believe I am testing `Not(A).canonical` in the test above (unless I'm misunderstanding your comment?) I really only added the `Not(e).canonical` for completeness, as this is the case that originally triggered the bug when  conversion to `A <= 1` was  suppressed by `with evaluate(False)`. I can remove the case `Not(e)` case, if you want?

smichr replied:
In the test, however, you don't use `evaluate=False` so you aren't dealing with an expression that is a Not. If you use `Not(A)`, you will be.

smichr replied:
But it might make more sense to do `return self(getattr(self.args[0], 'canonical', self.args[0]))`

mlange05 replied:
Fair enough. Removed.

mlange05 committed:
Logic: Removing unnecessary test for `Not(A > 1).canonical`

mlange05 commented:
@smichr Apologies if I misunderstood you, but I had already added a test `assert Not(A).canonical == Not(A)` to this PR and _that test is still there_. I only added (and now removed) the `Not(e).canonical == Not(e)`, as I wanted to make sure that `Relational`s are covered - but as you rightly pointed out, `Not(A > 1)` will be immediately simplified to `A <= 1`, so that second test is not adding anything new.

smichr commented:
> if I misunderstood you

OK, I see it now. What do you think about returning `self(getattr(self.args[0], 'canonical', self.args[0]))`?

mlange05 commented:
@smichr I tried it and got `TypeError: 'Not' object is not callable`. However, returning a new instance with `return Not(getattr(self.args[0], 'canonical', self.args[0]))` seems to work. Does that look alright to you?

smichr commented:
I left off the func: `return self.func(getattr(self.args[0], 'canonical', self.args[0]))`. This is preferred over returning Not since self may have been derived from Not (if I understand correctly...@asmeurer is the one that has pointed this out to me in the past).

mlange05 committed:
Logic: Better way to derive `.canonical` for `Not` instances

mlange05 commented:
@smichr Oh yes, sorry. I forgot that this was the sympy way to re-initialize things (PR updated). Thanks for pointing this out and all your patience.

oscargus commented:
I'm thinking that maybe it is better to add a property/method of `Relational` which simply negates the expression? `invert`? `negate`? Basically remapping the logic operator to the opposite.

A bit like `reversed`, but that actually invert it:
https://github.com/sympy/sympy/blob/f472bcc07e8af4080d0e78057dd8beb948d8766f/sympy/core/relational.py#L105-L125

I'm not sure what happens if to apply e.g. `And(x < 1, x >= 1)`, so
```
with evaluate(False):
    And(x < y, x >= y)
```
I think the purpose of the code is to detect this type of case. Maybe it still works (it evaluates to False without setting evaluate to False)? It seems like the code is maybe just applying a bit of "evaluation" when evaluate is false and that this is taken care of elsewhere when evaluate is true.

Out of curiosity, how did you stumble upon the bug?

smichr commented:
> I'm thinking that maybe it is better

I see what you mean:

```
>>> Not(And(x<0,y>1))
~((y > 1) & (x < 0))
>>> simplify(Not(And(x<0,y>1)))
~((y > 1) & (x < 0))
```
And rather than adding a canonical property to Not, perhaps it should be added to BooleanFunction to do what was done here on Not.

smichr committed:
move canonical to BooleanFunction

oscargus commented:
Yes, there may be cases where that can help. When working on the recent PRs, especially #15912, it became quite clear how important it is to be able to express identically meaning expressions on an identical format. So improving #15899 is quite crucial in some cases. If one can extend that to BooleanFunction it would of course be even better, so `Nand(a, b)` and `Or(Not(a), Not(b))` can be identified as the same expressions. Not obvious how to define the requirements there though, but a start is probably to not use Nand, Nor or Nxor in the canonical format, but to apply de Morgan.

smichr commented:
So shall we commit what is here?

oscargus commented:
Pros: this clarifies the need of a proper canonical for `BooleanFunction`s, the code is there and will solve a few issues

Cons: will there be proper canonicals for boolean functions?

The alternative as I see it is to add an `inverted` property as discussed above, That will solve all(?) usage in the code today (I've seen this at a few more places with the purpose of identifying the opposite relation so it should just be to replace `(~c).canonical` with `c.inverted.canonical`. At the moment I am more lending towards `inverted` as that does not mix `Boolean` and `Relational`. Although that mix seems quite natural from many aspects. Also, someone must write that property.

oscargus commented:
I have made an attempt at a property-based version for Relational in #15924. I think this may be a better way to solve this.

We'll see how it progresses, but if nothing else: thanks for bringing it up and suggesting a solution.

smichr commented:
> thanks for bringing it up and suggesting a solution.

ditto (and closing now).

smichr commented:
> better way to prevent the given example from triggering an `AttributeError`

The better way is just to know that you have to be careful when using attributes to make sure that the thing you computed has that attribute. So a change in the code where it is assumed that the result is a Relational would be appropriate: something before the following line should handle the non-Relational result

```
File "/base/data/home/apps/s~sympy-live-hrd/56.412575950366508090/sympy/sympy/logic/boolalg.py", line 538, in _new_args_filter
    nc = (~c).canonical
```

smichr committed:
watch for non-relational in _new_args_filter

smichr commented:
If the change I made looks ok, please delete the canonical property and add a test for the case you identified in the OP.

oscargus commented:
I think the only way where this will not be Relational is if Evaluate=False. It may be better to just skip the complete section then, if we so not even want this type of pre-checking (basically checks things like `And(Eq(x, y), Ne(x, y))`).

smichr committed:
Update test_boolalg.py

smichr committed:
remove BooleanFunction.canonical

The class shouldn't have to handle attributes that the arguments have since canonical for BooleanFunction could have different semantics than those of the arguments/Relationals.
